packages/mui-base/src/Dialog/Popup/DialogPopup.types.ts

```diff
@@ -0,0 +1,95 @@
+import * as React from 'react';
+import { type FloatingContext } from '@floating-ui/react';
+import { type BaseUIComponentProps } from '../../utils/BaseUI.types';
+import { DialogType, SoftCloseOptions } from '../Root/DialogRoot.types';
+
+export interface DialogPopupProps extends BaseUIComponentProps<'div', DialogPopupOwnerState> {
+  /**
+   * If `true`, the dialog supports CSS-based animations and transitions.
+   * It is kept in the DOM until the animation completes.
+   *
+   * @default false
+   */
+  animated?: boolean;
+  /**
+   * If `true`, the dialog element is kept in the DOM when closed.
+   *
+   * @default false
+   */
+  keepMounted?: boolean;
+}
+
+export interface DialogPopupOwnerState {
+  open: boolean;
+  modal: boolean;
+  nestedOpenDialogCount: number;
+}
+
+export interface UseDialogPopupParameters {
+  /**
+   * If `true`, the dialog supports CSS-based animations and transitions.
+   * It is kept in the DOM until the animation completes.
+   */
+  animated: boolean;
+  /**
+   * The id of the dialog element.
+   */
+  id?: string;
+  /**
+   * The ref to the dialog element.
+   */
+  ref: React.Ref<HTMLElement>;
+  /**
+   * Determines if the dialog is modal.
+   */
+  modal: boolean;
+  /**
+   * Determines if the dialog is open.
+   */
+  open: boolean;
+  /**
+   * Callback fired when the dialog is requested to be opened or closed.
+   */
+  onOpenChange: (open: boolean) => void;
+  /**
+   * The id of the title element associated with the dialog.
+   */
+  titleElementId: string | undefined;
+  /**
+   * The id of the description element associated with the dialog.
+   */
+  descriptionElementId: string | undefined;
+  /**
+   * Callback to set the id of the popup element.
+   */
+  setPopupElementId: (id: string | undefined) => void;
+  /**
+   * Type of the dialog (ordinary dialog or alert dialog).
+   */
+  type: DialogType;
+  /**
+   * Determines whether the dialog should close when clicking outside of it or pressing the escape key.
+   */
+  softClose: SoftCloseOptions;
```

Comment:
MDN docs call this `light dismiss` instead. I also think it should be the default if it isn't here

Reply:
I found a discussion in WHATWG about this. They are considering adding an attribute to control this to the native dialog element. The attribute names and values are not fully agreed upon yet, but it seems that `closedby=none|closerequest|any` is a likely option (https://github.com/whatwg/html/issues/9373#issuecomment-1960371221), with `closerequest` (= esc key) being the default.
We can mimic this behavior (and proposed attribute name) in our component.

Reply:
There seems to be a bunch of confusion/misunderstanding in that GH thread.

> I could see several types of modal dialog that really want explicit action. Like "Choose A or B" - cancel is not an option.

Any dialog that presents a choice between two options should be an `alertdialog`, which _must_ be modal, and also protect against clicking outside. So this is not relevant to the broader discussion around modality + protecting against clicking outside on modal dialogs.

When you separate dialogs that should be `alertdialog` role, I think there are very few cases where you want a modal dialog (but not an alertdialog), and also protect against clicking outside. The only examples I can think of are dialogs that contain a large form, or many configurable options like a checkbox group. It's not an alertdialog, but you want to protect against clicking outside, because unless the form data is in local storage, the data will be lost if the user closes the dialog accidentally.

But outside of those use cases, I think most modal dialogs should be easily dismissible. So I think the default modal dialog should be dismissible.

So we're left with the following dialog use cases:

1. Non-modal dialogs. For drawer/sheet components.
2. Dismissible Modal dialogs. This is the standard, default use case imo.
3. Modal dialogs that protect against clicking outside. For dialogs with large amounts of data. Should be opt-in imo.
4. Alert dialogs. For dialogs where a choice is presented that demands immediate attention.

Outside of a typical Drawer/Sheet component, I think the need for a non-modal dialog is questionable. Drawer/Sheet will need to be its own dedicated component anyway, because it needs to support swipe to dismiss functionality, which regular Dialogs shouldn't support. So unless we can think of some valid use cases for non-modal dialogs outside of Drawer/Sheet, we could consider dropping that functionality?

Mixing alertdialog with regular Dialog is problematic, because it allows the user to set both `alertdialog` and `dismissible` (or whatever problematic prop configuration we settle on) at the same time.

So we could do the following:

1. Drop support for non-modal dialog
2. Separate AlertDialog into its own dedicated component.
3. Just have a Dialog component that is always modal, and has a `dismissible` prop.

Thoughts?

Reply:
For the record, we decided to split the Dialog and AlertDialog, add the `dismissible` prop to the Dialog (both modal and non-modal), and always allow closing with the Esc key.